contracts/asset/src/execute/sale.rs

```diff
@@ -0,0 +1,109 @@
+use cosmwasm_std::{BankMsg, Coin, CustomMsg, DepsMut, Env, MessageInfo, Response};
+use cw721::traits::Cw721State;
+
+use crate::{error::ContractError, state::AssetConfig};
+
+pub fn buy<TNftExtension, TCustomResponseMsg>(
+    deps: DepsMut,
+    _env: &Env,
+    info: &MessageInfo,
+    id: String,
+    recipient: Option<String>,
+    deductions: Vec<(String, Coin, String)>,
+) -> Result<Response<TCustomResponseMsg>, ContractError>
+where
+    TNftExtension: Cw721State,
+    TCustomResponseMsg: CustomMsg,
+{
+    let asset_config = AssetConfig::<TNftExtension>::default();
+
+    let listing = asset_config
+        .listings
+        .may_load(deps.storage, &id)?
+        .ok_or(ContractError::ListingNotFound { id: id.clone() })?;
+
+    let price = listing.price.clone();
+    let seller = listing.seller.clone();
+
+    let mut payment = info
+        .funds
+        .iter()
+        .find(|coin| coin.denom == price.denom)
+        .ok_or_else(|| ContractError::NoPayment {})?
+        .clone();
```

Comment:
What happens if multiple coins are sent? Seems like everything other than the first match will be stuck in the contract. If only one token is expected, it should be enforced

Reply:
Adjusted to fail the tx if multiple coins are sent